File: content/tracks/dsa/problems/lc-0020-valid-parentheses/Solution.java
import java.util.*;

class Solution {
    public boolean isValid(String s) {
        // Each closing bracket must match the most recent unmatched opening bracket.
        Map<Character, Character> opener = Map.of(')', '(', ']', '[', '}', '{');
        Deque<Character> stack = new ArrayDeque<>();
        for (char ch : s.toCharArray()) {
            Character open = opener.get(ch);
            if (open != null) {
                if (stack.isEmpty() || !stack.peek().equals(open)) {
                    return false;
                }
                stack.pop();
            } else {
                stack.push(ch);
            }
        }
        // Anything left open was never closed.
        return stack.isEmpty();
    }
}

File: content/tracks/dsa/problems/lc-0020-valid-parentheses/solution.py
from typing import List


class Solution:
    def isValid(self, s: str) -> bool:
        # Each closing bracket must match the most recent unmatched opening bracket.
        opener = {")": "(", "]": "[", "}": "{"}
        stack: List[str] = []
        for ch in s:
            if ch in opener:
                if not stack or stack[-1] != opener[ch]:
                    return False
                stack.pop()
            else:
                stack.append(ch)
        # Anything left open was never closed.
        return not stack
